### ISSUE TITLE
heap corruption

### ISSUE THREAD

**Issue by robertswiecki** [google/kafel#2]
```
#define TCGETS          0x5401
#define PROT_NONE       0x0
#define PROT_READ       0x1
#define PROT_WRITE      0x2
#define PROT_EXEC       0x4

#define O_RDONLY        0x00000000
#define O_WRONLY        0x00000001
#define O_RDWR          0x00000002
#define O_CREAT         0x00000100
#define O_EXCL          0x00000200
#define O_NOCTTY        0x00000400
#define O_NONBLOCK      0x00004000
#define O_DIRECTORY     0x00200000
#define O_CLOEXEC       0x02000000
#define OPEN_MASK       0xFDDFBBFF

#define SAFE_WRONLY     0x00000301
#define SAFE_RDRW       0x00000302

#define SIGHUP           1
#define SIGINT           2
#define SIGQUIT          3
#define SIGILL           4
#define SIGTRAP          5
#define SIGABRT          6
#define SIGIOT           6
#define SIGBUS           7
#define SIGFPE           8
#define SIGKILL          9
#define SIGUSR1         10
#define SIGSEGV         11
#define SIGUSR2         12
#define SIGPIPE         13
#define SIGALRM         14
#define SIGTERM         15
#define SIGSTKFLT       16
#define SIGCHLD         17
#define SIGCONT         18
#define SIGSTOP         19
#define SIGTSTP         20
#define SIGTTIN         21
#define SIGTTOU         22
#define SIGURG          23
#define SIGXCPU         24
#define SIGXFSZ         25
#define SIGVTALRM       26
#define SIGPROF         27
#define SIGWINCH        28
#define SIGIO           29
#define SIGPOLL         29

POLICY AllowAllocations {
        ALLOW {
                brk,
                mmap,
                mprotect,
                munmap
        }
}

POLICY AllowBasicFsCalls {
        ALLOW {
                access,
                close {
                        fd > 2
                },
                getcwd,
                getdents,
                ioctl {
                        fd <= 2
                },
                lseek,
                newlstat,
                newfstat,
                newstat,
                open {
                        (flags == O_RDONLY) ||
                        (flags & OPEN_MASK == O_RDONLY) ||
                        (flags & OPEN_MASK == SAFE_WRONLY) ||
                        (flags & OPEN_MASK == SAFE_RDRW)
                },
                read,
                readlink,
                write {
                        fd > 2
                }
        }
}

POLICY AllowBasicIO {
        ALLOW {
                ioctl {
                        cmd == TCGETS
                },
                read {
                        fd == 0
                },
                readv {
                        fd == 0
                },
                write {
                        fd == 1 || fd == 2
                },
                writev {
                        fd == 1 || fd == 2
                }
        }
}

POLICY AllowIPC {
        ALLOW {
                pipe,
                pipe2
        }
}


POLICY AllowMisc {
        ALLOW {
                arch_prctl,
                exit_group,
                futex,
                getrlimit,
                set_robust_list,
                set_tid_address,
                sysinfo
        }
}

POLICY AllowSignals {
        ALLOW {
                rt_sigaction,
                rt_sigprocmask
        }
}

POLICY Main {
        ALLOW {
                execve
        },
        USE AllowAllocations,
        USE AllowBasicFsCalls,
        USE AllowBasicIO,
        USE AllowMisc,
        USE AllowSignals
}


USE Main DEFAULT KILL
```

**Comment by robertswiecki** [comment]
```
$ ./dump_policy_bpf </tmp/pol 
*** Error in `./dump_policy_bpf': malloc(): memory corruption: 0x000055fdfeb8fff0 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x790cb)[0x7f83045d30cb]
/lib/x86_64-linux-gnu/libc.so.6(+0x8421e)[0x7f83045de21e]
/lib/x86_64-linux-gnu/libc.so.6(__libc_calloc+0x27b)[0x7f83045e0c2b]
./dump_policy_bpf(+0x13a50)[0x55fdfcc43a50]
./dump_policy_bpf(+0x140cd)[0x55fdfcc440cd]
./dump_policy_bpf(+0x129d8)[0x55fdfcc429d8]
./dump_policy_bpf(+0x10f19)[0x55fdfcc40f19]
./dump_policy_bpf(+0x10563)[0x55fdfcc40563]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7f830457a3f1]
./dump_policy_bpf(+0x103aa)[0x55fdfcc403aa]
======= Memory map: ========
55fdfcc30000-55fdfcc4e000 r-xp 00000000 fd:00 1454983                    /home/jagger/src/nsjail/kafel/tools/dump_policy_bpf/dump_policy_bpf
55fdfce4d000-55fdfce61000 r--p 0001d000 fd:00 1454983                    /home/jagger/src/nsjail/kafel/tools/dump_policy_bpf/dump_policy_bpf
55fdfce61000-55fdfce62000 rw-p 00031000 fd:00 1454983                    /home/jagger/src/nsjail/kafel/tools/dump_policy_bpf/dump_policy_bpf
55fdfeb8d000-55fdfebae000 rw-p 00000000 00:00 0                          [heap]
7f8300000000-7f8300021000 rw-p 00000000 00:00 0 
7f8300021000-7f8304000000 ---p 00000000 00:00 0 
7f8304343000-7f8304359000 r-xp 00000000 fd:00 919447                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7f8304359000-7f8304558000 ---p 00016000 fd:00 919447                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7f8304558000-7f8304559000 r--p 00015000 fd:00 919447                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7f8304559000-7f830455a000 rw-p 00016000 fd:00 919447                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7f830455a000-7f8304717000 r-xp 00000000 fd:00 919887                     /lib/x86_64-linux-gnu/libc-2.24.so
7f8304717000-7f8304917000 ---p 001bd000 fd:00 919887                     /lib/x86_64-linux-gnu/libc-2.24.so
7f8304917000-7f830491b000 r--p 001bd000 fd:00 919887                     /lib/x86_64-linux-gnu/libc-2.24.so
7f830491b000-7f830491d000 rw-p 001c1000 fd:00 919887                     /lib/x86_64-linux-gnu/libc-2.24.so
7f830491d000-7f8304921000 rw-p 00000000 00:00 0 
7f8304921000-7f8304946000 r-xp 00000000 fd:00 919883                     /lib/x86_64-linux-gnu/ld-2.24.so
7f8304b15000-7f8304b17000 rw-p 00000000 00:00 0 
7f8304b41000-7f8304b45000 rw-p 00000000 00:00 0 
7f8304b45000-7f8304b46000 r--p 00024000 fd:00 919883                     /lib/x86_64-linux-gnu/ld-2.24.so
7f8304b46000-7f8304b47000 rw-p 00025000 fd:00 919883                     /lib/x86_64-linux-gnu/ld-2.24.so
7f8304b47000-7f8304b48000 rw-p 00000000 00:00 0 
7ffe6df63000-7ffe6df84000 rw-p 00000000 00:00 0                          [stack]
7ffe6dfc8000-7ffe6dfca000 r--p 00000000 00:00 0                          [vvar]
7ffe6dfca000-7ffe6dfcc000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
Aborted
```

And with ASAN

```
$ ./dump_policy_bpf </tmp/pol 
ASAN:DEADLYSIGNAL
=================================================================
==2844==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000514cc3 bp 0x0c3a000001d2 sp 0x7ffd070566f0 T0)
==2844==The signal is caused by a READ memory access.
==2844==Hint: address points to the zero page.
    #0 0x514cc2 in fix_tailq_moving /home/jagger/src/nsjail/kafel/range_rules.c:75:7
    #1 0x514cc2 in add_range_rule /home/jagger/src/nsjail/kafel/range_rules.c:97
    #2 0x5160d3 in add_missing_rules /home/jagger/src/nsjail/kafel/range_rules.c:254:7
    #3 0x5160d3 in normalize_rules /home/jagger/src/nsjail/kafel/range_rules.c:309
    #4 0x50ee61 in compile_policy /home/jagger/src/nsjail/kafel/codegen.c:551:3
    #5 0x50ca39 in kafel_compile /home/jagger/src/nsjail/kafel/kafel.c:106:10
    #6 0x50a765 in main /home/jagger/src/nsjail/kafel/tools/dump_policy_bpf/main.c:42:12
    #7 0x7f3fe1e4f3f0 in __libc_start_main /build/glibc-jxM2Ev/glibc-2.24/csu/../csu/libc-start.c:291
    #8 0x419679 in _start (/home/jagger/src/nsjail/kafel/tools/dump_policy_bpf/dump_policy_bpf+0x419679)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /home/jagger/src/nsjail/kafel/range_rules.c:75:7 in fix_tailq_moving
==2844==ABORTING
```

**Comment by happyCoder92** [comment]
Fixed with b59123e
